---
title: "Simulations"
author: "Sam Bogan"
date: '2023-01-27'
output: html_document
---

This markdown includes simulations of genetically-correlating gene expression traits and DNA methylation of corresponding gene bodies across 80 RNA-seq replicates derived froma 40 full-/half-sib families split and reared across two environments.

Current problem: I want to determine the effect of a continuous predictor on variation explained by a random effect. Presently, I can only do this by binning my methylation into quartiles. 


Spurs QG to do:

1. Measure and plot phenome wide distributions of VA and h2 by heritability
2. Look into 'family means' animal model
3. Simulate 100 independent traits
4. Add gaussian process for genetic correlation to model

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

setwd("~/Documents/GitHub/Methyl_QG_Spurp/")

# Load packages
library(tidyverse)
library(nadiv)
library(brms)
library(MASS)
library(holodeck)
library(bayestestR)
library(rstanarm)

```

Simulate data for one gene with genetic and environmental effects on DNA methylation and gene expression

```{r}

# Create experimental pedigree
pedi <- read.csv("pedigree.csv")
pedi_p <- prepPed(pedi)

#  Read in metadata
meta_df <- read.csv("pedigree_meta.csv")

# With nadiv, create a relatedness matrix
# Relatedness matrix is necessary for brms rather than inverse relatedness
relat_m <- as.matrix(makeA(pedi_p))

```

```{r}

# Simulate correlating correlations between maternal environment, methylation and gene expression
meta_df <- meta_df %>% 
  group_by(F0_env) %>% 
  sim_discr(n_vars = 2, var = 1, cov = 0.66, group_means = c(1, -1))

meta_df <- meta_df %>% rename("GE" = "V1",
                              "M" = "V2")

# Plot correlations
ggplot(data = meta_df,
       aes(x = M, y = GE)) +
  geom_smooth(method = "lm") +
  geom_point(aes(color = F0_env)) +
  theme_classic() +
  labs(color = "Maternal env", x = "Methylation", y = "Expression")

## To partition VA by methylation and differential methylation, creating quantiles for each
# Create ordinal methylation quantiles
meta_df <- meta_df %>%
    mutate(quantile = ntile(M, 4))

# Create ordinal diff meth quantiles


```

Fit brms animal using code from example by Julien Martin
https://juliengamartin.github.io/wam_tuto/brms-1.html

```{r}

# Fit multivariate model (GE and M) partitioning VA by methylation
f1 <- bf(GE ~ 1 + (1 | gr(ID, cov = relat_m, by = quantile)) + # Fits relatedness matrix w/ meth partitions for GE
       F0_env + F1_env + (1 | Dam) + (1 | Sire) + M) # Environ. + dam/sire effects on GE
f2 <- bf(M ~ 1 + (1 | gr(ID, cov = relat_m)) + # Fits relatedness matrix for M
         F0_env + F1_env + (1 | Dam) + (1 | Sire))

brms_m1.1 <- brm(f1 + f2 + set_rescor(FALSE), 
  data = meta_df,
  data2 = list(relat_m = relat_m),
  family = gaussian(),
  chains = 1, cores = 1, iter = 1000
)

```
#Effect of methylation on expression via genetic variance
Export $V_A$ then calculate and plot $h^2$

```{r}
# Estimate h2 grouped by methylation quantile
Var.table <- as_draws_df(brms_m1.1)

Var.table$h.bwt.1 <- as.mcmc((Var.table$`sd_ID__GE_Intercept:quantile1`)^2 / 
                               ((Var.table$`sd_ID__GE_Intercept:quantile1`)^2 + 
                                  (Var.table$sd_Dam__GE_Intercept)^2 + 
                                  (Var.table$sd_Sire__GE_Intercept)^2))

Var.table$h.bwt.2 <- as.mcmc((Var.table$`sd_ID__GE_Intercept:quantile2`)^2 / 
                               ((Var.table$`sd_ID__GE_Intercept:quantile2`)^2 + 
                                  (Var.table$sd_Dam__GE_Intercept)^2 + 
                                  (Var.table$sd_Sire__GE_Intercept)^2))

Var.table$h.bwt.3 <- as.mcmc((Var.table$`sd_ID__GE_Intercept:quantile3`)^2 / 
                               ((Var.table$`sd_ID__GE_Intercept:quantile3`)^2 + 
                                  (Var.table$sd_Dam__GE_Intercept)^2 + 
                                  (Var.table$sd_Sire__GE_Intercept)^2))

Var.table$h.bwt.4 <- as.mcmc((Var.table$`sd_ID__GE_Intercept:quantile4`)^2 / 
                               ((Var.table$`sd_ID__GE_Intercept:quantile4`)^2 + 
                                  (Var.table$sd_Dam__GE_Intercept)^2 + 
                                  (Var.table$sd_Sire__GE_Intercept)^2))

# Create df of h2 measures
h2.methyl <- as.data.frame(rbind(
  cbind(summary(Var.table$h.bwt.1)$statistics[1], summary(Var.table$h.bwt.1)$quantiles[1], summary(Var.table$h.bwt.1)$quantiles[5]),
  cbind(summary(Var.table$h.bwt.2)$statistics[1], summary(Var.table$h.bwt.2)$quantiles[1], summary(Var.table$h.bwt.2)$quantiles[5]),
  cbind(summary(Var.table$h.bwt.3)$statistics[1], summary(Var.table$h.bwt.3)$quantiles[1], summary(Var.table$h.bwt.3)$quantiles[5]),
  cbind(summary(Var.table$h.bwt.4)$statistics[1], summary(Var.table$h.bwt.4)$quantiles[1], summary(Var.table$h.bwt.4)$quantiles[5])
))

# Add methylation quantile to df as variable 
h2.methyl$quant <- c(1, 2, 3, 4)

# Plot h2 grouped by methylation quantile
ggplot(data = h2.methyl,
       aes(x = V1, y = quant, group = as.factor(quant), color = quant)) +
  geom_point(size = 3) +
  geom_errorbar(aes(xmin = V2, xmax = V3), width = 0) +
  theme_classic() +
  theme(legend.position = "none") +
  scale_color_viridis_c() +
  labs(y = "Methylation quartile", x = expression(italic("h")^2))
  
```

```{r}

# First, extract posterior samples
post <- posterior_samples(brms_m1.1)

# Hypothesis testing for differences in VA
bf <- bayesfactor_parameters(post$`sd_ID__GE_Intercept:quantile1`, 
                             post$`sd_ID__GE_Intercept:quantile4`,
                             direction = "two-sided", 
                             null = median(post$`sd_ID__GE_Intercept:quantile4`))

bf

plot(bf)
```


#Effect of environment on gene expression methylation
Input model results into mediation analysis

```{r}

# Calculate indirect effect by multiplying draws of mediator and predictor variables
post <-
  post %>% 
  mutate(ind = b_M_F0_envU * b_GE_M)

# Plot posterior distribution of indirect effect
post %>% 
  ggplot(aes(x = ind)) +
  geom_density(color = "black", fill = "skyblue") + 
  geom_vline(xintercept = 0, lty = 2, size = 1, color = "grey") +
  labs(y = "Density", x = "Mediation effect (posterior)") +
  theme_classic()

```

